File: tests/coretests/Stocks/Services/GapAnalysisTests.cs
using System.Collections.Generic;
using System.Linq;
using core.fs.Services;
using core.fs.Services.Analysis;
using coretests.testdata;
using Xunit;

namespace coretests.Stocks.Services;

public class GapAnalysisTests
{
    private readonly List<GapAnalysis.Gap> _gaps = GapAnalysis.detectGaps(
        TestDataGenerator.PriceBars(TestDataGenerator.NET),
        SingleBarPriceAnalysis.SingleBarAnalysisConstants.NumberOfDaysForRecentAnalysis
    );
    
    [Fact]
    public void NumberOfGapsDetected_Matches() => Assert.Equal(11, _gaps.Count);
    
    [Fact]
    public void Gap_PercentChange_Correct() => Assert.Equal(0.052m, _gaps[0].PercentChange, 3);
    
    [Fact]
    public void Gap_RelativeVolume_Correct() => Assert.Equal(0.94m, _gaps[0].RelativeVolume);
    
    [Fact]
    public void Gap_GapType_Correct() => Assert.Equal(GapAnalysis.GapType.Up, _gaps[0].Type);
    
    [Fact]
    public void ClosedQuickly_Matches() => Assert.Equal(7, _gaps.Count(g => g.ClosedQuickly));
    
    [Fact]
    public void Open_Matches() => Assert.Equal(4, _gaps.Count(g => g.Open));
}

File: tests/coretests.fs/Stocks/Services/GapAnalysisTests.fs
module coretests.fs.Stocks.Services.GapAnalysisTests

open Xunit
open core.fs.Services
open core.fs.Services.Analysis
open testutils
open FsUnit

let gaps =
    GapAnalysis.detectGaps
        (TestDataGenerator.PriceBars(TestDataGenerator.NET))
        SingleBarPriceAnalysis.SingleBarAnalysisConstants.NumberOfDaysForRecentAnalysis


[<Fact>]
let ``Number of Gaps Detected Matches`` () =
    gaps.Length |> should equal 11

[<Fact>]
let ``Gap PercentChange Correct`` () =
    System.Math.Round(gaps.[0].PercentChange, 3) |> should equal 0.052m

[<Fact>]
let ``Gap RelativeVolume Correct`` () =
    gaps.[0].RelativeVolume |> should equal 0.94m

[<Fact>]
let ``Gap GapType Correct`` () =
    gaps.[0].Type |> should equal GapAnalysis.GapType.Up

[<Fact(Skip = "Need to review if the logic is correct")>]
let ``ClosedQuickly Matches`` () =
    gaps |> Array.filter (fun g -> g.ClosedQuickly) |> Array.length |> should equal 7

[<Fact(Skip = "Need to review if the logic is correct")>]
let ``Open Matches`` () =
    gaps |> Array.filter (fun g -> g.Open) |> Array.length |> should equal 4